File: frontend/app/lib/profile-completeness.ts
import type { Profile } from './api';

function isEmpty(v: string | null | undefined): boolean {
  return !String(v ?? '').trim();
}

/**
 * Counts empty inputs required for a complete base resume (matches Edit profile form).
 * Used for dashboard badge and blocking tailored CV generation.
 */
export function getProfileCompleteness(profile: Profile): {
  emptyCount: number;
  paths: Set<string>;
} {
  const paths = new Set<string>();
  let emptyCount = 0;

  const mark = (path: string, empty: boolean) => {
    if (empty) {
      emptyCount += 1;
      paths.add(path);
    }
  };

  mark('basic.full_name', isEmpty(profile.full_name));
  mark('basic.headline', isEmpty(profile.headline));
  mark('basic.summary', isEmpty(profile.summary));
  mark('basic.email', isEmpty(profile.email));
  mark('basic.phone', isEmpty(profile.phone));
  mark('basic.address', isEmpty(profile.address));
  mark('basic.linkedin_url', isEmpty(profile.linkedin_url));
  mark('basic.photo', !String(profile.photo_base64 ?? '').trim());

  const exps = profile.experience || [];
  if (exps.length === 0) {
    emptyCount += 1;
    paths.add('experience.section');
  } else {
    exps.forEach((exp, i) => {
      mark(`experience.${i}.title`, isEmpty(exp.title));
      mark(`experience.${i}.company`, isEmpty(exp.company));
      mark(`experience.${i}.start_date`, isEmpty(exp.start_date));
      mark(`experience.${i}.end_date`, isEmpty(exp.end_date));
      mark(`experience.${i}.description`, isEmpty(exp.description));
    });
  }

  const edus = profile.education || [];
  if (edus.length === 0) {
    emptyCount += 1;
    paths.add('education.section');
  } else {
    edus.forEach((edu, i) => {
      mark(`education.${i}.school`, isEmpty(edu.school));
      mark(`education.${i}.degree`, isEmpty(edu.degree));
      mark(`education.${i}.field`, isEmpty(edu.field));
      mark(`education.${i}.start_date`, isEmpty(edu.start_date));
      mark(`education.${i}.end_date`, isEmpty(edu.end_date));
      mark(`education.${i}.description`, isEmpty(edu.description));
    });
  }

  const skills = profile.skills || [];
  const hasSkill = skills.some((s) => String(s).trim());
  mark('skills', !hasSkill);

  const certs = profile.certifications || [];
  certs.forEach((c, i) => {
    mark(`cert.${i}.name`, isEmpty(c.name));
    mark(`cert.${i}.authority`, isEmpty(c.authority));
    mark(`cert.${i}.date`, isEmpty(c.date));
    mark(`cert.${i}.url`, isEmpty(c.url));
  });

  const langs = profile.languages || [];
  if (langs.length === 0) {
    emptyCount += 1;
    paths.add('languages.section');
  } else {
    langs.forEach((lang, i) => {
      mark(`languages.${i}`, isEmpty(lang));
    });
  }

  return { emptyCount, paths };
}

export function countProfileEmptyFields(profile: Profile): number {
  return getProfileCompleteness(profile).emptyCount;
}

File: backend/app/profile_completeness.py
"""Count empty *required* profile fields (must match frontend app/lib/profile-completeness.ts)."""

from app.models import Profile


def _empty(v: str | None) -> bool:
    return not str(v or "").strip()


def count_required_empty_fields(profile: Profile) -> int:
    """Return number of missing required fields; 0 means profile is complete for generation."""
    n = 0

    if _empty(profile.full_name):
        n += 1
    if _empty(profile.summary):
        n += 1
    if _empty(profile.email):
        n += 1

    exps = profile.experience or []
    if len(exps) == 0:
        n += 1
    else:
        for exp in exps:
            if _empty(exp.title):
                n += 1
            if _empty(exp.company):
                n += 1
            if _empty(exp.start_date):
                n += 1
            if _empty(exp.end_date):
                n += 1

    edus = profile.education or []
    if len(edus) == 0:
        n += 1
    else:
        for edu in edus:
            if _empty(edu.school):
                n += 1
            if _empty(edu.degree):
                n += 1
            if _empty(edu.field):
                n += 1
            if _empty(edu.start_date):
                n += 1
            if _empty(edu.end_date):
                n += 1

    skills = profile.skills or []
    if not any(str(s).strip() for s in skills):
        n += 1

    for c in profile.certifications or []:
        if _empty(c.name):
            n += 1

    langs = profile.languages or []
    if len(langs) == 0:
        n += 1
    else:
        for lang in langs:
            if _empty(lang):
                n += 1

    return n


def profile_is_complete(profile: Profile) -> bool:
    return count_required_empty_fields(profile) == 0
